### ISSUE TITLE
macOS 10.13 - Comparing latest release (3.6.0) and current (4.0.0) - strange behavior

### ISSUE THREAD

**Issue by noaht8um** [hashcat/hashcat#1399]
I am seeing some strange behavior on macOS when comparing hashcat from homebrew and the latest git clone. And yes, I already know the state of OpenCL on macOS, and how hashcat is certainly not optimized for running on macOS, etc.

<details>
 <summary>Here are my devices:</summary>

```
/usr/local/bin/hashcat -I

OpenCL Info:

Platform ID #1
  Vendor  : Apple
  Name    : Apple
  Version : OpenCL 1.2 (Aug 23 2017 16:35:41)

  Device ID #1
    Type           : CPU
    Vendor ID      : 4
    Name           : Intel(R) Core(TM) i5-4670K CPU @ 3.40GHz
    Version        : OpenCL 1.2
    Processor(s)   : 4
    Clock          : 3400
    Memory         : 4096/16384 MB allocatable
    OpenCL Version : OpenCL C 1.2
    Driver Version : 1.1

  Device ID #2
    Type           : GPU
    Vendor ID      : 2147483648
    Vendor         : NVIDIA
    Name           : GeForce GTX 970
    Version        : OpenCL 1.2
    Processor(s)   : 13
    Clock          : 1080
    Memory         : 1024/4096 MB allocatable
    OpenCL Version : OpenCL C 1.2
    Driver Version : 10.25.28 378.10.10.10.15.120
```

</details>

When using "homebrew hashcat" (3.6.0) for WPA (-m 2500), I have been getting the "Checking for weak hashes" error but **only** when using a dictionary.

<details>
 <summary>("homebrew hashcat" Using Dictionary)</summary>

```
/usr/local/bin/hashcat -m 2500 test.hccapx 10_million_password_list_top_1000000.txt

hashcat () starting...

OpenCL Platform #1: Apple
=========================
* Device #1: Intel(R) Core(TM) i5-4670K CPU @ 3.40GHz, skipped.
* Device #2: GeForce GTX 970, 1024/4096 MB allocatable, 13MCU

Hashes: 2 digests; 1 unique digests, 1 unique salts
Bitmaps: 16 bits, 65536 entries, 0x0000ffff mask, 262144 bytes, 5/13 rotates
Rules: 1

Applicable optimizers:
* Zero-Byte
* Single-Hash
* Single-Salt
* Slow-Hash-SIMD

Watchdog: Temperature abort trigger disabled.
Watchdog: Temperature retain trigger disabled.

Checking for weak hashes...Abort trap: 6
```
</details>

<details>
 <summary>("homebrew hashcat" Using mask: '?d?d?d?d?d?d?d?d')</summary>

```
/usr/local/bin/hashcat -m 2500 -a 3 test.hccapx '?d?d?d?d?d?d?d?d'

hashcat () starting...

OpenCL Platform #1: Apple
=========================
* Device #1: Intel(R) Core(TM) i5-4670K CPU @ 3.40GHz, skipped.
* Device #2: GeForce GTX 970, 1024/4096 MB allocatable, 13MCU

Hashes: 2 digests; 1 unique digests, 1 unique salts
Bitmaps: 16 bits, 65536 entries, 0x0000ffff mask, 262144 bytes, 5/13 rotates

Applicable optimizers:
* Zero-Byte
* Single-Hash
* Single-Salt
* Brute-Force
* Slow-Hash-SIMD

Watchdog: Temperature abort trigger disabled.
Watchdog: Temperature retain trigger disabled.

[s]tatus [p]ause [r]esume [b]ypass [c]heckpoint [q]uit => s

Session..........: hashcat
Status...........: Running
Hash.Type........: WPA/WPA2
Hash.Target......: SSID (AP:00:00:00:00:00:00 STA:00:00:00:00:00:00)
Time.Started.....: Thu Oct 19 13:54:54 2017 (5 secs)
Time.Estimated...: Thu Oct 19 14:06:33 2017 (11 mins, 34 secs)
Guess.Mask.......: ?d?d?d?d?d?d?d?d [8]
Guess.Queue......: 1/1 (100.00%)
Speed.Dev.#2.....:   143.2 kH/s (10.87ms)
Recovered........: 0/1 (0.00%) Digests, 0/1 (0.00%) Salts
Progress.........: 585728/100000000 (0.59%)
Rejected.........: 0/585728 (0.00%)
Restore.Point....: 0/10000000 (0.00%)
Candidates.#2....: 42345678 -> 41327234

[s]tatus [p]ause [r]esume [b]ypass [c]heckpoint [q]uit =>
```

</details>

I know the error was closed as a known issue in [#1350](https://github.com/hashcat/hashcat/issues/1350), but I saw it was fixed in 4.0.0. After cloning and making the new binary, I ran it to find that it did indeed fix the error:

<details>
  <summary>(hashcat 4.0.0 Using Dictionary)</summary>

```
~/Downloads/hashcat/hashcat -m 2500 test.hccapx 10_million_password_list_top_1000000.txt

hashcat (4.0.0-rc6) starting...

OpenCL Platform #1: Apple
=========================
* Device #1: Intel(R) Core(TM) i5-4670K CPU @ 3.40GHz, skipped.
* Device #2: GeForce GTX 970, 1024/4096 MB allocatable, 13MCU

Hashes: 2 digests; 1 unique digests, 1 unique salts
Bitmaps: 16 bits, 65536 entries, 0x0000ffff mask, 262144 bytes, 5/13 rotates
Rules: 1

Applicable optimizers:
* Zero-Byte
* Single-Hash
* Single-Salt
* Slow-Hash-SIMD-LOOP

Password length minimum: 8
Password length maximum: 63

Watchdog: Temperature abort trigger disabled.
Watchdog: Temperature retain trigger disabled.

Dictionary cache hit:
* Filename..: 10_million_password_list_top_1000000.txt
* Passwords.: 999999
* Bytes.....: 8529113
* Keyspace..: 999999

[s]tatus [p]ause [r]esume [b]ypass [c]heckpoint [q]uit => s

Session..........: hashcat
Status...........: Running
Hash.Type........: WPA/WPA2
Hash.Target......: SSID (AP:00:00:00:00:00:00 STA:00:00:00:00:00:00)
Time.Started.....: Thu Oct 19 14:08:56 2017 (3 secs)
Time.Estimated...: Thu Oct 19 14:09:19 2017 (20 secs)
Guess.Base.......: File (10_million_password_list_top_1000000.txt)
Guess.Queue......: 1/1 (100.00%)
Speed.Dev.#2.....:    35972 H/s (9.96ms)
Recovered........: 0/1 (0.00%) Digests, 0/1 (0.00%) Salts
Progress.........: 275667/999999 (27.57%)
Rejected.........: 175827/275667 (63.78%)
Restore.Point....: 255944/999999 (25.59%)
Candidates.#2....: turnleft -> Letmein4

[s]tatus [p]ause [r]esume [b]ypass [c]heckpoint [q]uit =>
```
</details>

<details>
  <summary>(hashcat 4.0.0 Using mask: '?d?d?d?d?d?d?d?d')</summary>

  ```
  ~/Downloads/hashcat/hashcat -m 2500 -a 3 test.hccapx '?d?d?d?d?d?d?d?d'

  hashcat (4.0.0-rc6) starting...

  OpenCL Platform #1: Apple
  =========================
  * Device #1: Intel(R) Core(TM) i5-4670K CPU @ 3.40GHz, skipped.
  * Device #2: GeForce GTX 970, 1024/4096 MB allocatable, 13MCU

  Hashes: 2 digests; 1 unique digests, 1 unique salts
  Bitmaps: 16 bits, 65536 entries, 0x0000ffff mask, 262144 bytes, 5/13 rotates

  Applicable optimizers:
  * Zero-Byte
  * Single-Hash
  * Single-Salt
  * Brute-Force
  * Slow-Hash-SIMD-LOOP

  Password length minimum: 8
  Password length maximum: 63

  Watchdog: Temperature abort trigger disabled.
  Watchdog: Temperature retain trigger disabled.

  [s]tatus [p]ause [r]esume [b]ypass [c]heckpoint [q]uit => s

  Session..........: hashcat
  Status...........: Running
  Hash.Type........: WPA/WPA2
  Hash.Target......: NETGEAR68 (AP:10:da:43:dd:9d:08 STA:a4:8d:3b:ae:3a:a4)
  Time.Started.....: Thu Oct 19 14:15:55 2017 (1 sec)
  Time.Estimated...: Thu Oct 19 15:01:48 2017 (45 mins, 52 secs)
  Guess.Mask.......: ?d?d?d?d?d?d?d?d [8]
  Guess.Queue......: 1/1 (100.00%)
  Speed.Dev.#2.....:    36322 H/s (10.58ms)
  Recovered........: 0/1 (0.00%) Digests, 0/1 (0.00%) Salts
  Progress.........: 26624/100000000 (0.03%)
  Rejected.........: 0/26624 (0.00%)
  Restore.Point....: 0/10000000 (0.00%)
  Candidates.#2....: 22345678 -> 21223123

  [s]tatus [p]ause [r]esume [b]ypass [c]heckpoint [q]uit =>
  ```

</details>

However, the speed is ***much*** slower. As the dictionary attack failed on "homebrew hashcat" I can only compare the mask attacks.

I get this from "homebrew hashcat":
`Speed.Dev.#2.....:   143.2 kH/s (10.87ms)`
And this from the new one:
`Speed.Dev.#2.....:    36322 H/s (10.58ms)`

I tried using a larger mask, smaller mask, changing to the hashcat make directory, etc. but nothing would increase the speed. As dictionary cracking is the most viable method (IMHO) for cracking WPA passwords, this does make things sucky for Mac users.

I also tried this on my 2013 MacBook with a dedicated GPU and the newest version ran slower as well (interestingly, only slower on the dedicated NVIDIA GPU; the integrated Iris graphics stayed the same speed), just with a less dramatic speed hit (about 4x slower).

If anyone has any advice (other than using another OS) I'd love to hear it and/or test it.

**TL;DR** Current hashcat build fixes `Abort trap: 6` error but is about 30x slower the latest release on macOS.

P.S. I used collapsible code blocks because I knew I'd be posting a crapload of terminal pastes, but if it affects readability (couldn't figure out how to make a newline after a code block), I can re-post this issue without them.

**Comment by jsteube** [comment]
Can you please retry with -w 4 (this enables 256 threads instead of 32). Also make sure to use the latest github version, I changed it several times today.

**Comment by noaht8um** [comment]
I just tried it again with the latest GitHub version (`git pull; make clean; make`) and I went from about 35000 H/s to 40000 H/s using -w 4. Unfortunately this is still significantly less than the old version.

**Comment by jsteube** [comment]
Then it's unknown what causes this. We need to interactively debug this. Please come to IRC freenode channel #hashcat and PM me (atom).

**Comment by jsteube** [comment]
I've eventually found the issue and added some code to handle it with commit https://github.com/hashcat/hashcat/commit/2e85972ec55191d9050988571d3de29b8653e456

Please retry your test

**Comment by jsteube** [comment]
please test and close the issue if fixed

**Comment by noaht8um** [comment]
@jsteube Thanks for fixing this! Works great now. Tested on my hackintosh and my MacBook Pro with an NVIDIA card.
